### PR TITLE
addEdge check - endpoints exist

### LINKED ISSUE #50: ExpressionGraph.addEdge(): check that both endpoints are in the graph

### PR DESCRIPTION
**Pull request by maxqchen**
addEdge check - endpoints exist

Fixes #50

### CODE REVIEW

**Comment by maxqchen** [comment]
Oh wow Intellij's create pull request thing worked. That's so cool.

**Comment by maxqchen** [comment]
Bah checkstyles. I thought we were at 120 characters? (or maybe that was just in core). Shall I change this to 120 as well?

File: checkstyle.xml [hunk 1]
```diff
@@ -145,7 +145,7 @@ page at http://checkstyle.sourceforge.net/config.html -->
 
     <module name="LineLength">
       <!-- Checks if a line is too long. -->
-      <property name="max" value="${com.puppycrawl.tools.checkstyle.checks.sizes.LineLength.max}" default="80"/>
+      <property name="max" value="${com.puppycrawl.tools.checkstyle.checks.sizes.LineLength.max}" default="120"/>
```
Comment (lucaswoj) [inline checkstyle.xml]: I, and the general programming community, are against this, even though its sorta inconvenient sometimes. http://stackoverflow.com/questions/903754/do-you-still-limit-line-length-in-code

File: src/main/java/org/manifold/compiler/front/ExpressionGraph.java [hunk 2]
```diff
@@ -52,17 +49,18 @@ public void addVertex(VariableIdentifier vID)
...
-    return vertices;
+  public Collection<ExpressionVertex> getVertices() {
+    return allVertices;
   }
 
   private List<ExpressionEdge> edges = new ArrayList<>();
   public void addEdge(ExpressionEdge e) {
+    Preconditions.checkArgument(getVertices().contains(e.getSource())
+                    && e.getTarget() == null || getVertices().contains(e.getTarget()),
+            "Edge had unexpected vertices " + e.toString());
```
Comment (lucaswoj) [inline src/main/java/org/manifold/compiler/front/ExpressionGraph.java]: The indentation here is sorta confusing. I'd probably write this as

```
Preconditions.checkArgument(
    (e.getTarget() == null && getVertices().contains(e.getSource())) || getVertices().contains(e.getTarget()),
     "Edge had unexpected vertices " + e.toString()
);
```

Comment (lucaswoj) [inline src/main/java/org/manifold/compiler/front/ExpressionGraph.java]: Under what conditions would the target be `null`?

File: src/test/java/org/manifold/compiler/front/TestExpressionGraph.java [hunk 3]
```diff
@@ -1,5 +1,22 @@
...
+    expressionGraph = new ExpressionGraph();
+  }
 
+  @Test(expected = IllegalArgumentException.class)
+  public void testAddEdge_verticesNotInGraph_exceptionThrown() {
+    ExpressionVertex from = new ConstantValueVertex(expressionGraph, BooleanValue.getInstance(true));
+    ExpressionVertex to = new ConstantValueVertex(expressionGraph, BooleanValue.getInstance(false));
+    ExpressionEdge edgeOfTomCruise = new ExpressionEdge(from, to);
+    expressionGraph.addEdge(edgeOfTomCruise);
+  }
```
Comment (lucaswoj) [inline src/test/java/org/manifold/compiler/front/TestExpressionGraph.java]: It seems like there are a couple more cases in your code. To get complete branch coverage you should test

1. When the target is null
2. When the target is not in the graph
2. When the source is not in the graph

**Comment by lucaswoj** [comment]
Almost there! The changes shouldn't take more than 10 mins.

File: src/main/java/org/manifold/compiler/front/ExpressionGraph.java (hunk 2 above)
Comment (maxqchen) [inline src/main/java/org/manifold/compiler/front/ExpressionGraph.java]: There are cases where the to vertex is null initially and added later (check the graph builder code).

As for the indentation, both were 1 level deeper than they should've been, not sure why. The 1st line checks the source, and the 2nd line checks the target. The 1st line could be on its own line but that seems unnecessary and increases vertical space used...

Also that 2nd line you wrote is 112 characters long :stuck_out_tongue_closed_eyes:

Comment (lucaswoj) [inline src/main/java/org/manifold/compiler/front/ExpressionGraph.java]: Vertical space is free!

Comment (maxqchen) [inline src/main/java/org/manifold/compiler/front/ExpressionGraph.java]: Arguably free, though. Less wasted vertical space = more of a function / functions displayed on screen = don't need to scroll up/down as much. This is getting too nitty, and tbh I don't really care too much for it. New diff incoming, I'll put it on its own line.

**Comment by mtrberzi** [comment]
Hey Max! We talked about the correct cases for this offline, so once the test cases are there, I'll review this again.

**Comment by maxqchen** [comment]
Do you mean the tests lucas mentioned? There is a new diff that added them.

File: src/main/java/org/manifold/compiler/front/ExpressionGraph.java [hunk 4]
```diff
@@ -52,17 +49,19 @@ public void addVertex(VariableIdentifier vID)
...
-    return vertices;
+  public Collection<ExpressionVertex> getVertices() {
+    return allVertices;
   }
 
   private List<ExpressionEdge> edges = new ArrayList<>();
   public void addEdge(ExpressionEdge e) {
+    Preconditions.checkArgument(
+        getVertices().contains(e.getSource())
+            && e.getTarget() == null || getVertices().contains(e.getTarget()),
```
Comment (TysonAndre) [inline src/main/java/org/manifold/compiler/front/ExpressionGraph.java]: `A && B || C` is parsed left to right as `(A && B) || C`. I think you mean 
` && (e.getTarget() == null || getVertices().contains(e.getTarget()))`

**Comment by mtrberzi** [comment]
Looks good OK to ship!

**Pull request merged**